HousingFinanceInterimApi/V1/Gateways/LogParserGateway.cs

```diff
@@ -0,0 +1,121 @@
+using Amazon.CloudWatchLogs.Model;
+using HousingFinanceInterimApi.V1.Gateway.Interfaces;
+using HousingFinanceInterimApi.V1.Infrastructure;
+using HousingFinanceInterimApi.V1.Handlers;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using System.Linq;
+
+namespace HousingFinanceInterimApi.V1.Gateways
+{
+    public class LogParserGateway : ILogParserGateway
+    {
+        private readonly IDatabaseContext _context;
+
+        public LogParserGateway(IDatabaseContext context)
+        {
+            _context = context;
+        }
+
+        public async Task UpdateDatabaseWithResults(string logGroupName, List<List<ResultField>> queryResults)
+        {
+            // Validate input parameters
+            if (string.IsNullOrWhiteSpace(logGroupName))
+            {
+                throw new ArgumentNullException(nameof(logGroupName), "Log group name cannot be null or empty.");
+            }
+
+            // Handle null or empty queryResults
+            if (queryResults is null || !queryResults.Any())
+            {
+                await LogResultAsync(logGroupName, DateTime.UtcNow, true).ConfigureAwait(false);
+                return;
+            }
+
+            try
+            {
+                foreach (var result in queryResults)
+                {
+                    var logEntry = ProcessResult(logGroupName, result);
+
+                    // If a failure is found, log it and stop processing further results
+                    if (logEntry != null && !logEntry.IsSuccess)
+                    {
+                        await LogResultAsync(logEntry).ConfigureAwait(false);
+                        return;
+                    }
+                }
+
+                // If no failures were found, log a success entry
+                await LogResultAsync(logGroupName, DateTime.UtcNow, true).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                LoggingHandler.LogError($"Unexpected error in UpdateDatabaseWithResults for log group '{logGroupName}': {ex.Message}");
+                LoggingHandler.LogError(ex.StackTrace);
+                throw;
+            }
+        }
+
+        private NightlyProcessLog ProcessResult(string logGroupName, List<ResultField> result)
+        {
+            try
+            {
+                var timestamp = result.Find(r => r.Field == "@timestamp")?.Value;
+                var message = result.Find(r => r.Field == "@message")?.Value;
+
+                if (DateTime.TryParse(timestamp, out var parsedTimestamp))
+                {
+                    var isSuccess = message is not null && !message.Contains("error", StringComparison.OrdinalIgnoreCase);
```

Comment:
I don't follow this.
My initial image of this based on the comment above the CloudWatch text query was that the query is meant to look for failures directly. But here we're looking for failure within the returned query results.
I'm guessing this will change after dev testing?

Reply:
For each Cloudwatch Log Group, The `LogParserUseCase` class filters for the word "error". If any of the returned messages contains this string, it will pass these messages to the `LogParserGateway` class - shown here.

If there are any returned messages containing "error", `LogParserGateway` will save the Log Group name, the timestamp and the IsSuccess as False (failed) once, and move on to the next LogGroup Name. If no messages are returned by CW this signifies that there were no errors and the Log Group Name is saved with the isSuccess flag as True (succeeded).

This is done to ensure that we save the Log Group once per call.